modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotResponseRemoteFutureTask.java

```diff
@@ -72,70 +74,65 @@ public SnapshotResponseRemoteFutureTask(
+            List<SnapshotMetadata> metas = cctx.snapshotMgr().readSnapshotMetadatas(snpName, snpPath);
 
-            for (Map.Entry<Integer, Set<Integer>> e : parts.entrySet()) {
-                ofNullable(e.getValue()).orElse(Collections.emptySet())
-                    .forEach(p -> handled.add(new GroupPartitionId(e.getKey(), p)));
-            }
+            Function<GroupPartitionId, SnapshotMetadata> findMeta = pair -> {
+                for (SnapshotMetadata meta : metas) {
+                    Map<Integer, Set<Integer>> parts0 = meta.partitions();
 
-            snpSndr.init(handled.size());
+                    if (F.isEmpty(parts0))
+                        continue;
 
-            File snpDir = cctx.snapshotMgr().snapshotLocalDir(snpName, snpPath);
+                    Set<Integer> locParts = parts0.get(pair.getGroupId());
 
-            List<CompletableFuture<Void>> futs = new ArrayList<>();
-            List<SnapshotMetadata> metas = cctx.snapshotMgr().readSnapshotMetadatas(snpName, snpPath);
+                    if (locParts != null && locParts.contains(pair.getPartitionId()))
+                        return meta;
+                }
 
-            for (SnapshotMetadata meta : metas) {
-                Map<Integer, Set<Integer>> parts0 = meta.partitions();
+                return null;
+            };
 
-                if (F.isEmpty(parts0))
-                    continue;
+            Map<GroupPartitionId, SnapshotMetadata> partsToSend = new HashMap<>();
 
-                handled.removeIf(gp -> {
-                    if (ofNullable(parts0.get(gp.getGroupId()))
-                        .orElse(Collections.emptySet())
-                        .contains(gp.getPartitionId())
-                    ) {
-                        futs.add(CompletableFuture.runAsync(() -> {
-                            if (err.get() != null)
-                                return;
+            for (Map.Entry<Integer, Set<Integer>> e : parts.entrySet())
+                e.getValue().forEach(p -> partsToSend.computeIfAbsent(new GroupPartitionId(e.getKey(), p), findMeta));
 
-                            File cacheDir = cacheDirectory(new File(snpDir, databaseRelativePath(meta.folderName())),
-                                gp.getGroupId());
+            if (partsToSend.containsValue(null)) {
+                Collection<GroupPartitionId> missed = F.viewReadOnly(partsToSend.entrySet(), Map.Entry::getKey,
+                    e -> e.getValue() == null);
 
-                            if (cacheDir == null) {
-                                throw new IgniteException("Cache directory not found [snpName=" + snpName + ", meta=" + meta +
-                                    ", pair=" + gp + ']');
-                            }
+                err.compareAndSet(null, new IgniteException("Snapshot partitions missed on local node " +
+                    "[snpName=" + snpName + ", missed=" + missed + ']'));
```

Comment:
I assume that we should also stop executing this method at this point.(

Reply:
Additionally, we can move this check higher and remove redundant data re-scan.

Reply:
> I assume that we should also stop executing this method at this point.
Done, thanks